### ISSUE TITLE
[HOLD for payment 2023-10-23] [$500] Login -  App allows changing to use the 2FA code option when signning in with recovery code

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#28254]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
- User has set up 2FA
1. Go to the Login page
2. Enter email of a user that already set up 2FA
3. Enter a correct validation code
4. On verifying 2FA code, click "Use recovery code" option
5. Enter incorrect recovery code with enough 8 chars and click "Sign in" button
6. While "Sign in" button is showing the waiting icon, click "Use two-factor authentication code" option

## Expected Result:
 App should not allow changing to use the 2FA code option if we're verifying with the "recovery code" option.  The "Use two-factor authentication code" should be disabled during this state

## Actual Result:
 App allows changing to use the 2FA code option

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.74.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/34c346c7-a560-4b96-b221-449fbcccb343

https://github.com/Expensify/App/assets/93399543/87eeaac9-7006-4d9d-8c42-c9b65504bd03

https://github.com/Expensify/App/assets/93399543/71318f28-905b-47c9-a8af-8984c62bc909

https://github.com/Expensify/App/assets/93399543/f1814088-25c9-4041-b414-d8f31f95931e

https://github.com/Expensify/App/assets/93399543/5ccc2c74-3d6a-4a27-848c-d46e5cceec24

https://github.com/Expensify/App/assets/93399543/2f0d6091-c0d6-4f5a-b6d7-7fc3c7a36d40

https://github.com/Expensify/App/assets/93399543/e276face-46a8-4048-b044-56730a82f04c

**Expensify/Expensify Issue URL:**
**Issue reported by:**  @hoangzinh
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1695420775197399

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013c221635a4525e47</li>
        <li>Upwork Job ID: 1706643652898590720</li>
        <li>Last Price Increase: 2023-09-26</li>
<li>Automatic offers: </li>
<ul>
<li>hoangzinh | Contributor | 27102881</li>
</ul></li>
    </ul>
</details>

**Comment by kbecciv** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
App allows changing to use the 2FA code option when Signing in with recovery code
### What is the root cause of that problem?
Currently, we haven't disabled the text button used to switch between 2FA and Recovery code in any cases https://github.com/Expensify/App/blob/df6e2b67259d521d5aacca2f5b2fffc7e6a17f77/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js#L326-L336
That's the reason why the Sign-in form is loading but the above text button still is clickable
### What changes do you think we should make in order to solve the problem?
We need to pass a `disabled` prop to the above text button with the value same as the loading state for Sign-in button here https://github.com/Expensify/App/blob/df6e2b67259d521d5aacca2f5b2fffc7e6a17f77/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js#L385
We can put it into a variable so we can reuse it in 2 places above.

**Comment by saranshbalyan-1234** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
The "Use two-factor authentication code"  is clickable even during the api call is going on for signin request
### What is the root cause of that problem?
The "Use two-factor authentication code" test is wrapped in  PressableWithFeedback,  which also expect disabled as prop but we have not passed any value.
### What changes do you think we should make in order to solve the problem?
https://github.com/Expensify/App/blob/44568ba24c3ac00b3a43d07d81a8061f687dadba/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js#L356-L369
inside PressableWithFeedback we need to pass props.account.isLoading  that's it
The Solution would look something like this
```javascript
<PressableWithFeedback
disabled={props.account.isLoading}
``` 
### What alternative solutions did you explore? (Optional)
N/A

**Comment by Shivam7-1** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The problem is that the app allows users to switch to the "Use two-factor authentication code" option even when they are in the process of verifying their identity using a recovery code. This behavior is not desirable, as it can lead to confusion and errors.
### What is the root cause of that problem?
The root cause of this problem is that the "Use two-factor authentication code" option is not properly disabled when the user is verifying their identity with a recovery code.
### What changes do you think we should make in order to solve the problem?

**Problem Description:**

The problem is that the app allows users to switch to the "Use two-factor authentication code" option even when they are in the process of verifying their identity using a recovery code. This behavior is not desirable, as it can lead to confusion and errors.

**Root Cause:**

The root cause of this problem is that the "Use two-factor authentication code" option is not properly disabled when the user is verifying their identity with a recovery code.
The Problem occurs in this code file https://github.com/Expensify/App/blob/main/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js

**Proposed Solution:**

To solve this problem, Here I Compiled Certain Steps Parts where we should update the code to disable the "Use two-factor authentication code" option when the user is verifying with a recovery code. 
Here are the proposed changes:

1. **Add a State Variable for Verification Method:**

   We should add a state variable `verificationMethod` to keep track of whether the user is verifying with a recovery code or a two-factor authentication code. Initialize it to `'2FA'` by default.

   ```
   const [verificationMethod, setVerificationMethod] = useState('2FA'); 
   ```

2. **Update the `switchBetween2faAndRecoveryCode` Function:**

   Modify the `switchBetween2faAndRecoveryCode` function to toggle the `verificationMethod` state and clear input fields and form errors accordingly.

   ```
   const switchBetween2faAndRecoveryCode = () => {
       // Toggle the verification method
       setVerificationMethod((prevMethod) => (prevMethod === '2FA' ? 'recoveryCode' : '2FA'));

       // Clear inputs and errors
       setRecoveryCode('');
       setTwoFactorAuthCode('');
       setFormError({});

       if (props.account.errors) {
           Session.clearAccountMessages();
       }
   };
   ```

3. **Update the Render Logic:**

   In the render logic, disable the "Use two-factor authentication code" option when the `verificationMethod` is `'recoveryCode'`.

   
 ```
  <PressableWithFeedback
       style={[styles.mt2]}
       onPress={switchBetween2faAndRecoveryCode}
       underlayColor={themeColors.componentBG}
       hoverDimmingValue={1}
       pressDimmingValue={0.2}
       accessibilityRole={CONST.ACCESSIBILITY_ROLE.BUTTON}
       accessibilityLabel={verificationMethod === 'recoveryCode' ? props.translate('recoveryCodeForm.use2fa') : props.translate('recoveryCodeForm.useRecoveryCode')}
       disabled={verificationMethod === 'recoveryCode'} // Disable when verifying with recovery code
   >
       <Text style={[styles.link]}>{verificationMethod === 'recoveryCode' ? props.translate('recoveryCodeForm.use2fa') : props.translate('recoveryCodeForm.useRecoveryCode')}</Text>
 
   </PressableWithFeedback>
   
```

In Conclusion For above Proposal

The proposed solution involves adding a state variable `verificationMethod` to keep track of the verification method, updating the `switchBetween2faAndRecoveryCode` function to toggle this state, and disabling the "Use two-factor authentication code" option when verifying with a recovery code. These changes should ensure that users cannot switch to the 2FA option while in the process of verifying with a recovery code, addressing the problem described.

### What alternative solutions did you explore? (Optional)
NA

**Comment by sonialiap** [comment]
@robertKozik what do you think of the proposals?

**Comment by robertKozik** [comment]
Hi all! Thank you for your proposals! I suggest selecting @hoangzinh's proposal, as it was the first one posted and correctly identified the root cause and Pressable element to patch. I believe disabling the link based on the loading state is sufficient; introducing an additional state seems redundant in this case.

Selected proposal: https://github.com/Expensify/App/issues/28254#issuecomment-1735427519

🎀 👀 🎀 C+ reviewed

**Comment by sonialiap** [comment]
@Li357 your proposal was selected 🎉  please proceed with a PR. If you have one, please link it to this issue

**Comment by Li357** [comment]
@hoangzinh's proposal looks good, assigning!

**Comment by sonialiap** [comment]
I'm OOO Oct 16-23. Adding leave buddy

@peterdbarkerUK hoangzinh requires payment on Oct 23 for reporting and fixing the issue. I'm not sure what the multiple hold messages in the title mean, maybe regressions?

**Comment by peterdbarkerUK** [comment]
Yep, I've seen a couple like this and will take care of it tomorrow/Tuesday :)

**Comment by peterdbarkerUK** [comment]
@hoangzinh could you complete this checklist and notify me here so that I can complete the pay out?

**Comment by robertKozik** [comment]
Yeah, that should be on my end as C+ here
> * [X]  [@hoangzinh / @robertKozik] The PR that introduced the bug has been identified. Link to the PR: It was an edge case which wasn't discovered before, no PR introduced it in particular.
> * [X]  [@hoangzinh / @robertKozik] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: **N/A**
> * [X]  [@hoangzinh / @robertKozik] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **N/A**
> * [X]  [@hoangzinh / @robertKozik] Determine if we should create a regression test for this bug. As this bug didn't broke any user flow, I think we shouldn't create a regression test.
> * [X]  [@hoangzinh / @robertKozik] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again. **N/A**

**Comment by sonialiap** [comment]
Thanks Robert! Back from OOO so grabbing this back from Pete to complete the payments now ⏳

**Comment by sonialiap** [comment]
@hoangzinh fix + bonus = $750 - paid ✔️ 
@hoangzinh report = $50 - sent the above payment before checking reporter 🤦 sent a new offer for the report!

**Comment by sonialiap** [comment]
@hoangzinh please accept the offer on Upwork for reporting this issue :)

**Comment by sonialiap** [comment]
All paid ✅
